Add a string replace method to handle multiple search keys efficiently.

shana opened the pull request:
There are multiple places where we want to replace multiple characters or strings with one character (e.g. cleaning up paths), and calling replace for every source string is wasteful, we can do it all in one go.

_Contributed with ❤️ by W4Games_

shana committed:
Add a string replace method to handle multiple search keys efficiently.

There are multiple places where we want to replace multiple characters or strings
with one character (e.g. cleaning up paths), and calling replace for every source
string is wasteful, we can do it all in one go.

lyuma commented on core/math/color.cpp:
```diff
@@ -415,11 +415,7 @@ Color Color::named(const String &p_name, const Color &p_default) {
-	name = name.replace("_", "");
-	name = name.replace("'", "");
-	name = name.replace(".", "");
+	name = name.replace(Vector<String>({ " ", "-", "_", "'", "." }), "");
```
Does it make sense to use static storage for these vectors to avoid the allocation overhead each time, such as
```suggestion
	static Vector<String> to_replace = { " ", "-", "_", "'", "." };
	name = name.replace(to_replace, "");
```

shana replied:
Oh yes, probably!

YeldhamDev commented:
Would be nice if this was exposed to the script API.

shana committed:
Fix off-by-one error and add a string replace test

shana committed:
Make string vectors to replace static so we're not allocating every time

shana committed:
Expose `replace(Vector<String>, String)` to script as `replace_all`

shana commented:
Ok, I fixed the off-by one error [insert meme here], moved things to static storage, and exposed the replace overload as `replace_all` to script (which feels logical to me, open to suggestions on the naming)

shana committed:
Missed a spot

aaronp64 commented on core/string/ustring.cpp:
```diff
@@ -4093,6 +4093,43 @@ String String::replacen(const char *p_key, const char *p_with) const {
+
+		for (const String &key : p_keys) {
+			const int key_len = key.length();
+			if (src_pos + key_len > len)
```
```suggestion
			if (src_pos + key_len > len || key_len == 0)
```

If `p_keys` contains an empty string, `key_len` being zero could cause the `src_pos += key_len - 1;` line below to set `src_pos` back one, cancelling out the loop increment and repeatedly adding `p_with`

dalexeev commented on doc/classes/String.xml:
```diff
@@ -744,6 +744,14 @@
 				Replaces all occurrences of [param what] inside the string with the given [param forwhat].
 			</description>
 		</method>
+		<method name="replace_all" qualifiers="const">
```
In my opinion, the name `replace_multiple` would be better, since otherwise the user might think that `replace_all()` is a version of `replace()`. But `replace()` already replaces all occurrences, not just the first one (like in some languages/libraries).

We should also think about naming if we decide to add a pair replacement method in the future (`replace_*(["a", "b"], ["x", "y"])` or `replace_*({"a": "x", "b": "y"})`).

AThousandShips replied:
Might also be good to clarify that it does not replace recursively, i.e. `"banana".replace_all(["b","na"],"x")` gives `"xaxx"` not `"xxx"`

shana committed:
Make sure empty key strings are handled correctly.

Co-authored-by: Aaron Pagano <136198872+aaronp64@users.noreply.github.com>

dalexeev commented:
The efficiency looks obvious, but it would still be nice to do benchmarks if you don't mind.

Also, in my opinion, a separate method would be better than an overload, since it reduces the differences between the core and the scripting API. It's easier for new contributors and search-and-replace changes.

AThousandShips commented on core/string/ustring.cpp:
```diff
@@ -4093,6 +4093,43 @@ String String::replacen(const char *p_key, const char *p_with) const {
+
+		for (const String &key : p_keys) {
+			const int key_len = key.length();
+			if (src_pos + key_len > len || key_len == 0)
+				continue;
```
```suggestion
			if (src_pos + key_len > len || key_len == 0) {
				continue;
			}
```
Always use brackets

AThousandShips commented:
I agree on benchmarks primarily to aid in deciding *where* to use these methods, as the convenience of this is great, but important to not carelessly use it if it is slower (especially on input where only a few keys are present, or the keys are long, etc.)

AThousandShips commented on core/math/color.cpp:
```diff
@@ -415,11 +415,8 @@ Color Color::named(const String &p_name, const Color &p_default) {
-	name = name.replace("'", "");
-	name = name.replace(".", "");
+	static Vector<String> to_replace({ " ", "-", "_", "'", "." });
+	name = name.replace(to_replace, "");
```
Since several of these are plain erasures of individual characters I wonder if we should add an erase method as well, or even instead, that takes a vector of `char32_t` instead, it would be faster and cleaner I'd say

AThousandShips replied:
I'll see about writing a PR for these two options later

shana committed:
Update ustring.cpp

Co-authored-by: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>

shana commented:
> I'm not sure that the benchmark is as applicable to Godot, since it does not use STL and the results may differ for the better or for the worse. I meant comparison for different source string lengths, number and length of keys, etc. But I don't insist on benchmarks if it's tedious for you.

This benchmark is using the two versions of replace and calling them. I copy pasted the code directly, and the only difference is that the String class is the std implementation and not ours, but none of that matters for the algorithm performance. The core bulk of the logic is our code in both cases, only the `find` and `substr` implementations are the ones from std instead of our versions - and I'm pretty sure ours are not a magnitude faster than the ones in std, so the difference really doesn't matter.

What you see in that benchmark is the result of a replace of a bunch of keys on a *very* short string - you can run the benchmark yourself and put in an actual long string, and you'll see the massive difference the size of the input string makes to the overall performance of calling replace a bunch of times.

lyuma commented:
What if we defer the decision to expose the method to scripting (and the name bikeshedding), and keep this as an internal core performance optimization?

GDScript users already can use [RexEx.sub](https://docs.godotengine.org/en/stable/classes/class_regex.html) if they need multiple things replaced in one pass.

clayjohn commented:
> What if we defer the decision to expose the method to scripting (and the name bikeshedding), and keep this as an internal core performance optimization?
> 
> GDScript users already can use [RexEx.sub](https://docs.godotengine.org/en/stable/classes/class_regex.html) if they need multiple things replaced in one pass.

Yes please. Let's not block performance optimizations so that we can bikeshed scripting API naming conventions. 

As for the internal naming. I can see there being a slight benefit to having a unique name per overload. But this function already has overloads, so I don't think it is the right time to be establishing a new naming convention

aaronp64 commented:
If the main goal of this change is performance optimization, I think it makes sense to benchmark using Godot's `String` class, ideally with different combinations of long/short keys, long/short original `String`, no/many replacements, etc.  I think the use of `CowData` in Godot's `String`, and many `std::string` implementations using small string optimization, makes them different enough that they could have performance differences here (might not matter, but would be nice to know for sure).

Even with the `std::string` benchmark at https://quick-bench.com/q/oUdTdeQHp1PN1kGWefV6tXw0eJk, `ReplaceAll` seems to be slower with larger `from` strings compared to `ReplaceOneAtATime`.  Changing "hello" to "hello hello hello hello hello hello hello hello hello hello" in both methods is enough to make `ReplaceOneAtATime` faster.

AThousandShips commented:
For benchmarking I'd say it's important to compare replacing strings with almost all the sequences searched for being empty, compared to repeated replacements, there might be slowdowns because of the way it works there though I haven't tested, but it's a possible edge case where performance might be hurting.

clayjohn commented:
It seems like the big difference with big datasets is that in the new approach handles not finding any matches worse. So a source string like:
```
": Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type"
```
Is much faster with the new method than with the old. 

But inserting lorem ipsum in the middle makes it slower than the old method
```
: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.ModedsdsLorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type: Shader.Modedsds : VisualShader.Type
```

Aaaaaand replacing String's find method with Godot's find method makes the new method take the lead again: https://quick-bench.com/q/uWnljUL9PkwmapILxlTTgJMGshM

It seems that Godot's find method is waaaaay slower than std's. 

_Original Benchmark with the lorem ipsum text from above:_
![Screenshot from 2024-05-28 18-30-49](https://github.com/godotengine/godot/assets/16521339/4215797d-e0f7-4627-9982-7f384eafd93c)

_Same as above, but using Godot's find method:_
![Screenshot from 2024-05-28 18-31-05](https://github.com/godotengine/godot/assets/16521339/348ee904-f083-4cfc-baed-0e936aad9ccf)

AThousandShips commented:
> It seems like the big difference with big datasets is that in the new approach handles not finding any matches worse.

This is what I suspected, I think optimization for searches would be best served by adding better underlying search mechanisms, and in some critical places using RegEx when possible etc.

Using a proper searcher, like a Knuth–Morris–Pratt or other pre-processed methods, when applicable would probably do the most for improvement

The problem is that the approaches are pretty brute force, and struggles for partial matches etc.

This feature is great for usability, but to solve performance the underlying mechanisms need improving IMO

shana commented:
@clayjohn Super interesting finds, thanks for digging into that!

I see other PRs happening around this, so feel free to close this one if there's going to be an equivalent but different improvement incoming from other PRs.

> as the input data is entirely unknown

Not necessarily completely unknown; for example, when sanitizing filepaths, we know the input data is small, so the impact of no matches when doing a replace all is negligible, but the memory impact is equal or better, since we're doing a lot of keys and therefore have more likelihood of multiple matches.

clayjohn commented:
Two more quick benchmarks

1. using the lorem ipsum (long string, few matches) with only single character keys

<img width="703" alt="Screenshot 2024-05-30 at 10 58 03 AM" src="https://github.com/godotengine/godot/assets/16521339/d8fe8032-7ba5-49ed-93d0-7941b83554df">

2. Using a shorter string with some matches "shader_name=clay+some-other*awesome--combo" with only single character keys

<img width="697" alt="Screenshot 2024-05-30 at 11 01 47 AM" src="https://github.com/godotengine/godot/assets/16521339/84165e64-096e-4497-a9cb-34febd1655cb">

I think we can extract some general principles from the benchmarking above:
1. This PR tends to be better than vanilla replace when the source string is short
2. This PR is much better than vanilla replace when there are lots of matches
3. This PR scales worse than vanilla replace when the string is long and there are few matches

So I think we could safely use it all all cases where we have a sense that the source string is small (filepaths, variable names, etc.). But we should avoid it for long strings (shader source, script source, etc.)

Mickeon commented:
The optimization of https://github.com/godotengine/godot/pull/92546 may have made the speed difference between `replace()` and `replace_all()` a bit too close. Just speculation on my part.
